Resiliency to newly-added capabilities

logiclrd opened the pull request:
Backblaze periodically adds new "capabilities" to the B2 API. The current strategy for parsing the models associated with successful logins requires all capabilities to be recognized. This breaks the library every time it happens. This PR aims to provide resiliency to this situation, by making the raw parsing of the "Allowed" structure in new type `AllowedRaw` just use a list of strings for the capabilities, and then later sorting them into known and unknown capabilities for the actual `Allowed` object returned by the `Connect` call.

logiclrd committed:
Added property UnknownCapabilities to the Allowed.cs class. Updated DebuggerDisplay to represent unknown capabilities in the returned string.
Made a copy of the Allowed class called AllowedRaw.cs in Bytewizer.Backblaze.Client.Internal that represents capabilities as a list of plain strings, with method ParseCapabilities to transform instances into the original Allowed class. Updated LangVersion in Backblaze.Client.csproj to 8.0 to allow the use of the coalescing assignment operator.
Made a copy of AuthorizedAccountResponse.cs in Bytewizer.Backblaze.Client.Internal that uses AllowedRaw instead of Allowed.
Updated AuthorizeAccountAsync to use AuthorizeAccountResponseRaw for the actual request and then convert the result to the regular AuthorizedAccountResponse structure, with ParseCapabilities splitting off unknown capabilities.

microcompiler commented:
@logiclrd Thanks again for your contribution.   I will take a look that these pull requests this weekend.   You don't by any chance have any test cases for these additions?

logiclrd commented:
I tested them by hand, by temporally undoing the previous PR that added the new capabilities for bucket notifications. I'll have to look into whether there's a simple way to test this from the other side...

logiclrd committed:
Added automated testing of Capabilities Resiliency:
- Added an AssemblyInfo class to Backblaze.Client.csproj with an InternalsVisibleTo attribute allowing Backblaze.Tests.Unit access to internal members.
- Added references from Backblaze.Tests.Unit.csproj to NSubstitute, Bogus, FluentAssertions and ASP.NET Core (in order to use Kestrel).
- Added class TestServer.cs to Backblaze.Tests.Unit to create an in-process hosted web server for testing purposes.
- Added text fixture class ApiClientFixture.cs to Backblaze.Tests.Unit with tests of the AuthorizeAccountAync method, including specifically Capabilities Resiliency.
Moved the initialization of Allowed.Capabilities and Allowed.UnknownCapabilities out of the loop in AllowedRaw.ParseCapabilities and added explicit initialization for Capabilities.

logiclrd committed:
Changed test AuthorizeAccountAync_should_parse_unknown_capabilities to use explicit strings instead of Faker-generated strings for unknown capability names, since there is a tiny change of faker.Random.Words(3) returning the same string twice.

logiclrd commented:
@microcompiler There you go. :-)

logiclrd committed:
Moved TestServer.cs and ApiClientFixture.cs into the correct namespace for the Backblaze.Tests.Unit project.

khvatskyi commented:
Any updates on this one?
